transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown.adoc

```diff
@@ -0,0 +1,118 @@
+
+NOTE: Heuristic transactions cannot be resolved automatically thus they are left untouched during suspension and then they will be reported to the administrator when Narayana gets resumed.
+
+_In cloud environments_, when hosting entities (e.g. containers, pods, virtual machines) are scaled down and their state is erased (e.g. their file system, ip address, and memory are deleted) WildFly cannot guarantee that all transactions will be completed and a data integrity issue might occur.
+
+NOTE: There are additional complications when hosting entities crash (i.e. an unintentional shutdown occurs). This extreme scenario is beyond what WildFly can handle and the hosting cloud environment must provide the ability to recover the hosting entity from crashes. In this case, the state of the hosting entity (e.g. its file system, ip address, etc.) should be recovered by the cloud environment without the intervention of the administrator. In K8s, for example, the employment of StatefulSet creates a stable environment where crashes are handled automatically. If a particular cloud environment does not provide a guarantee similar to StatefulSet, it is not possible to employ WildFly’s Transaction Subsystem. *As there are no solutions to solve this extreme scenario within WildFly (as it is impossible to control the lifecycle of the host from WildFly), this proposal is only concerned with problem 1.1.*
+
+== Issue Metadata
+
+=== Issue
+
+* [WFLY-17742](https://issues.jboss.org/browse/WFLY-17742)
+
+=== Dev Contacts
+
+* mailto:{jfinelli@redhat.com}[{Manuel Finelli}]
+
+=== QE Contacts
+
+* mailto:{istraka@redhat.com}[{Ivan Straka}]
+
+=== Testing By
+
+* [ ] Engineering
+* [X] QE
+
+=== Affected Projects or Components
+
+* [wildfly-core](https://github.com/wildfly/wildfly-core)
+* [wildfly](https://github.com/wildfly/wildfly)
+* [narayana](https://github.com/jbosstm/narayana)
+
+=== Relevant Installation Types
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* This proposal is about importing into WildFly the transactional functionalities of its Kubernetes Operator. It does not include any further improvements. Please, for further information, refer to the `transaction_recovery` controller in the Kubernetes Operator of WildFly.
+* The Management Model of WildFly will not be modified
+* The Transactions subsystem needs to become aware of in-doubt and heuristic transactions and delay its suspension until all in-doubt transactions are complete and all transactions that ended with a heuristic outcome have been resolved. This will guarantee data integrity in all scenarios, including situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
```

Comment:
So is this paragraph stating that the transaction subsystem will not return from suspend until there are no in doubt or heuristically completed transactions left in the store? As it stands the paragraph is somewhat unclear, particularly in the case where the timeout is non-negative.

Reply:
The paragraph says "it [the transaction subsystem] will delay its suspension until all in-doubt transactions are complete and all transactions that ended with a heuristic outcome have been resolved" but that will not happen if the timeout is positive since a positive timeout means the server will shutdown when the time limit is reached.

BTW I have already approved the proposal, these comments are my caveats.

Reply:
Got it! I will include the positive timeout into this point. Thanks :-)